### PR TITLE
feat(typography): create typography component FE-3068, FE-3002

### PR DESCRIPTION
**Pull request by jamime**
### Proposed behaviour

- Create `Typography`,`List` and `ListItem` components.
- Refactor palette so all colors can be used by the component directly.
- Add `color` util which can be used by all components.
- Update `box` to use new `color` util.
- Add MDX documentation for `Typography` and `Color`.
- Add eslint rule to prevent users using styled-system color directly.
- Add missing `backgroundColor` to Color definition.

> **BREAKING CHANGE:** The `atOpacity` function has been removed from the theme palette. To use this utility you should import it directly `import atOpacity from “carbon-react/lib/style/utils/at-opacity`.

> **BREKAING CHANGE:** The file “carbon-react/lib/style/index” has been removed.
If you want to use `addOpacity` you can use `atOpacity` directly.
If you want to use `blackAtOpacity` you can use `theme.palette.blackOpacity` or you can use `atOpacity` directly.
Please note that `blackAtOpacity` previously returned a white opacity you can use `theme.palette.whiteOpacity` to keep this behaviour.
If you want to use `generatePalette` you can import it directly `import generatePalette from “carbon-react/lib/style/palette`.

image

### Current behaviour

- No typography components
- Duplicated code around `palette` and `atOpacity`
- Missing `backgroundColor` in Color definition.

### Checklist

- [x] Commits follow our style guide
- [x] Screenshots are included in the PR if useful
- [x] All themes are supported if required
- [x] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
- [x] Storybook added or updated if required
- [x] Typescript `d.ts` file added or updated if required
- [x] Carbon implementation and Design System documentation are congruent

### Additional context
FE-3068, FE-3002

### Testing instructions

See the new MDX pages for Color and Typography. No need for cypress tests, there are extensive unit tests.

### CODE REVIEW

**Review by nicktitchmarsh — APPROVED** [review]

File: docs/colors.stories.mdx [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import LinkTo from '@storybook/addon-links/react';
+
+
+<Meta
+  title="Documentation/Colors"
+/>
+
+# Colors
+
+Some componets such as <LinkTo kind='Design System/Typography'>Typography</LinkTo> have a `color`, `bg`/`backgroundColor` and `opacity` props, this functionality is provided by <a href="https://styled-system.com/api/#color">`styled-system/color`</a>.
```
Comment (sjohnson-sage) [inline docs/colors.stories.mdx:10]:
```diff
components
componets
```

File: src/components/typography/list.component.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+);
+
+List.propTypes = {
+  ...Typography.propTypes,
+  as: propTypes.oneOf(['ul', 'ol'])
+};
+
+const ListItem = ({ children, ...props }) => (
+  <Typography
+    as='li' variant='p'
```
Comment (sjohnson-sage) [inline src/components/typography/list.component.js:29]: Really minor but I'd put these props on separate lines as above. I'll hurry up and get prettier in next pioneer Friday so we don't even have to think about this!

File: src/components/typography/typography.component.js [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+  variant, as, fontSize, fontWeight, textTransform, lineHeight, textDecoration
+}) => {
+  return {
+    as: as || getAs(variant),
+    size: fontSize || getSize(variant),
+    weight: fontWeight || getWeight(variant),
+    textTransform: textTransform || getTransform(variant),
+    textDecoration: textDecoration || getDecoration(variant),
+    lineHeight: lineHeight || getLineHeight(variant),
+    defaultMargin: variant === 'p' ? '0 0 16px 0' : '0'
```
Comment (sjohnson-sage) [inline src/components/typography/typography.component.js:139]: Minor again but you don't actually need the last 0 for the margin here
```suggestion
    defaultMargin: variant === 'p' ? '0 0 16px' : '0'
```

That will set left and right margins to 0

**Review by sjohnson-sage — COMMENTED** [review]
Looks really good, just a few pretty minor comments from me then happy to approve 👍

**Review by sjohnson-sage — APPROVED** [review]

File: docs/colors.stories.mdx [hunk 4]
```diff
@@ -0,0 +1,50 @@
+import LinkTo from '@storybook/addon-links/react';
+
```
Comment (DlgSHi) [inline docs/colors.stories.mdx:2]: unnecessary empty line :)

File: docs/colors.stories.mdx [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+```jsx
+<Typography color='red'>...</Typography>
+<Typography color='#ff0000'>...</Typography>
+<Typography color='rgba(100, 0, 0, 1)'>...</Typography>
+```
+
+## Palette Colors
+
+Using the name of a palette color ensures if we update the theme your components will also be updated, for this reason we advise not using a CSS sting.
```
Comment (mkrds) [inline docs/colors.stories.mdx:21]:
```suggestion
Using the name of a palette color ensures if we update the theme your components will also be updated, for this reason we advise not using a CSS string.
```

**Review by sjohnson-sage — APPROVED** [review]

**Review by DlgSHi — DISMISSED** [review]

**Review by nicktitchmarsh — APPROVED** [review]

**Comment by jamime** [comment]
@harpalsingh found some missing props from the variant enum in the manual table. I'll add them.

**Comment by jamime** [comment]
I've added the missing props from the variant enum in the manual table. I also added `backgroundColor` to the typescript definitions and to the styled-system prop table.

**Review by nicktitchmarsh — APPROVED** [review]

**Review by DlgSHi — APPROVED** [review]

**Review by sjohnson-sage — APPROVED** [review]

**Review by mkrds — APPROVED** [review]

**Pull request merged**